### DISCUSSION TITLE
Hide Rich Text View ???

### DISCUSSION THREAD

**Discussion by jbk68** [cortexrd/Knack-Toolkit-Library#117] [General]
Using KTL, I'm trying to figure out how to hide a Rich Text View based on another fields value using a KTL keyword.

Example: hide Rich Text if field_100 = "YES".

Any and all help would be appreciated... JON

**Comment by CSWinnall** [comment]
Hi Jon

_cfv should do this as long as it's on the view view qith the field you are checking against.  The options value of cfv is h.

`_cfv=[field_xxx, is, YES,,,,h], [ktlTarget, $('#view_yyy')]`

This is untested but should work

Craig

**Comment by jbk68** [comment]
Hey Craig, that worked.  Not sure if it's possible, but I'd like to use the keyword _ar (my favorite) to auto refresh the rich text view field.  

What I'm noticing is that when my referenced field (#255) changes from YES to NO, the rich text view won't always hide/show based on field #255 unless I refresh browser and I don't know where I can put the _ar keyword on the rich text view.  Do you ???

Thanks again for your help... JON

**Comment by CSWinnall** [comment]
If you could show me a screenshot of the page.  How is the yes updated?

Craig

**Comment by jbk68** [comment]
I was simply going into the builder and toggling (updating) the value for field 255 from YES to NO, NO to YES, to see if the suggestion you posted worked.  Field 255 on the table is changing as I make changes,  but I have the keyword _ar set to 5 so that the view/table updates every 5 seconds to reflect any data changes.

Just looking for a way to have rich text view (view_276) update as well using same keyword and settings.

Just wondering if there is a way to place the KTL keywords on a rich text view.

**Reply by cortexrd** [reply]
Jon,

Currently, only _oln and _ols are supported in Rich Text views.  That's just because they don't have titles or descriptions like all others.
But I'd like to add more keywords support for sure.

I'll see with Charles if we can do this in the parser, so all keywords would automatically be supported, just like many other view types.

Norm

**Comment by CSWinnall** [comment]
Knacks page rules will do this so that what ever the value when some comes
to the page it will either be hidden or not.

If you need the rich text view to hide based on the YES/NO in the table you
would need the table to ar not the rich text view.

**Comment by jbk68** [comment]
Hmm..., currently when I manually change field's data from "NO" to "YES" , the rich text hides **AND** the field in table updates to "YES" seamlessly (no refresh needed) thanks to _ar keyword.  But when I manually change field's data from "YES" to "NO", the rich text DOESN'T show on view but the field in the table now shows "NO" again seamlessly thanks to _ar.  In order for me to show the rich text at this point, I have to either go to another page and then back or refresh browser (chrome).

It's no big deal if I get it working or not. I was simply trying to show a message on the screen based on field's value.  And the only way I know how to show a message is by using a rich text view.  It shows up, but I wish there was a way to refresh the rich text view like you can a grid or details view using _ar.

Thanks again... JON

**Comment by CSWinnall** [comment]
Hi 

In looking at the ktl.js the _ar keyword will re-render the table and will also run any code on the view as well that specifically a `knack-view-render`.  To get the rich text view top hide you may have to use code as I am not sure if the keywords will re-run on an _ar. 

A question for Norm

Craig

**Comment by cortexrd** [comment]
Hi guys,

First, I think the reason why the Rich Text doesn't always hide/show as expected is because is because all views are refreshed asynchronously, (random order).

The best way to achieve your goal is to have the _ar on the view that has your field_100 (seen as the trigger), then also add _rvr=richtext view, so they are linked, but in a proper and predictable order.

_rvr is Refresh View after Refresh.

Norm

**Comment by CSWinnall** [comment]
Thank you Norm

I'd forgotten about that.

Craig

**Reply by jbk68** [reply]
Hey Norm, thanks for the reply.  Tried your suggestion and I'm still having an issue of the rich text view not hiding / showing properly.

Here is what I did: _ar=5 _rvr=view_253 _cfv=[field_376, is, NO,,,,h], [ktlTarget, $('#view_253')] 

This is placed on the view that contains field_376.  When I manually change field 376 from NO to YES, rich text hides as expected, but when I change field 376 from YES to NO, rich text view stays hidden unless I refresh browser or change to another page then back.

Thanks... JON

**Reply by cortexrd** [reply]
Jon,

It would be so much easier if we could see the content of your page.  Can you upload a screen grab (fake or blurred data is ok), so we can better understand your use case?

I'm having a hard time grasping what are the fields, belonging to what view, each view type, etc.

If I have to modify the KTL, I'll do it, but I'm sure that what you want to do is easy, and doable as is.

Norm

**Comment by cortexrd** [comment]
Guys,

I found the problem!  It was a pretty tough one.

I know it's pretty technical, but in short, it is due to the **ktlDisplayNone** class that is added by the _cls keyword when the condition is met (YES in this case).  

That class remains active after a hide command from the _cls because by design, there is no way to remove it using the _cls keyword.

Instead, we must explicitly remove the class using the _cls keyword like this:

`_cls=[!ktlDisplayNone], [ktlTarget, view_id]`

To summarize everything, you should have this in your driving view.  I assume it's a Details view:

Assuming that the rich text view is view_253

```
_ar=5
_cls=[!ktlDisplayNone], [ktlTarget, view_253]
_rvr=view_253
_cfv=[field_376, is, NO,,,,h], [ktlTarget, view_253)]
```

The _cls line will remove the ktlDisplayNone class, showing the view_253.
Then the _cfv will hide it back if condition is met.

Notes:
1. you may briefly see the view_253.  This is something we'll eventually fix, with some optimization.
2. when using ktlTarget, you don't need a jQuery selector with the full $('#view_352') syntax, thanks to the universal selector.  Just the view Id will do when it's something simple like this.  Actually, both will work just fine.
3. I need to update the KTL because it will not work as is now.  Give me a couple of hours, I'll let you know when it's ready.

Norm

**Reply by jbk68** [reply]
Thanks Norm, much appreciated... JON

**Comment by cortexrd** [comment]
Jon,

Can you try the beta version please?  I can't do a full release at this point due to code still in development.

The version **0.16.4-beta** is the one you need.

Your loader portion in the Javascript pane should look like this:

```
KnackInitAsync = function ($, callback) { // Load the Knack Toolkit Library (KTL)
    (window.LazyLoad = LazyLoad) && LazyLoad.js(['https://ctrnd.s3.amazonaws.com/Lib/KTL/KTL_Start.js'], () => {
        loadKtl($, callback, (typeof KnackApp === 'function' ? KnackApp : null), '0.16.4-beta' /*KTL version, leave blank to get latest*/, 'full'/*min or full*/);
    })
};

//... rest of your code goes here...
```
Then once this is added, you can try the fix above with the _cls keyword.

Norm

**Reply by jbk68** [reply]
Hey Norm, that worked!!! I manually changed field 376 from YES to NO and the back to NO to YES, and the rich text view was hidden / shown properly.

I did notice in the above fix on line 4 (_cfv line) that I had to correct the following to make it work.

_cfv=[field_376, is, NO,,,,h], [ktlTarget, view_253)]    **to**   _cfv=[field_376, is, NO,,,,h], [ktlTarget, $('#view_253')]

I noticed before your fix and after, the rich text when hidden is briefly showing when you go back to the page with view 253, and I belief you said you will fix with some optimization.  

Thanks again for the fix... JON

**Reply by cortexrd** [reply]
Jon,

You're right about this:

> I did notice in the above fix on line 4 (_cfv line) that I had to correct the following to make it work.
_cfv=[field_376, is, NO,,,,h], [ktlTarget, view_253)] to _cfv=[field_376, is, NO,,,,h], [ktlTarget, **$('#view_253')**]

But this has been fixed in the official (non beta) release.  You can now use the view_id directly.

Norm

**Comment by cortexrd** [comment]
Hi Jon,

I've just completed a major KTL update again - we're now at **0.17.0**.  I will update the documentation shortly, but here's the solution for the Rich Text view you need to hide, based on a condition.

Rich Text views now support keywords, and `_hv` now support `ktlCond`.

So add this in your Rich Text view:

`_hv=ktlCond, is, NO, fieldName, viewName`

It means, hide this view if fieldName in viewName is NO.
You can also remove this form above:  `_cls=[!ktlDisplayNone], [ktlTarget, view_253]` since not needed anymore.

Let me know if this works as expected.
Norm

**Comment by jbk68** [comment]
Hey Norm, I tried adding the _hv to the rich text view and it didn't work.  Not 100% where to place the keyword on the rich text, but as you know, there is a lot of options due to the fact it doesn't have a title field or description field.  

I placed it where the existing text was, both below and above existing text and in the html part both with unsuccessfully results.

Here is my keywords I tried: _hv=ktlCond, is, NO, field_376, view_120  **and I tried** _hv=ktlCond, is, NO, AH, Customer Info

Thank again Norm...

**Reply by cortexrd** [reply]
Hmmm... this should have worked, both actually.
The kw placement is the same as all others: **at the end**, i.e. right **after** what you intend to keep.
Are you sure you have version **KTL v0.17.0** ?  Ctrl+F5 to bypass cache.

Send us a screen grab showing the view IDs.  You can use the **KTL Developer Tools** popup to enable this:  https://github.com/cortexrd/Knack-Toolkit-Library/discussions/47#discussioncomment-6807182

Norm

**Comment by jbk68** [comment]
Hey Norm, I'm a 110% sure I'm using 17.0.  I placed the keyword at the end of the my text in rich text.  

So basically it looks like this in the Rich Text View:  MY TEST TEXT _hv=ktlCond, is, NO, field_376, view_120

Still can't get it to hide when value changes from NO to YES.  

Field 376 is a field from a connected table, would that matter?  Field 376 is on and view able in the table in view 120.  When I manually change the data in builder, I can see the changes for field 376 in view 120 made instantly.

Thanks for the Sunday morning help... JON

**Reply by jbk68** [reply]
Hey Norm, any thoughts on why my rich text (view_253) is not hiding based on the above post?

**Reply by cortexrd** [reply]
Jon,

You description is a bit confusing.

You say:
`_hv=ktlCond, is, NO, field_376, view_120`
Still can't get it to hide when value changes from **NO to YES.**

This looks backwards to me.  Your code says:  **hide rich text view when NO**.

What is **view_120** - what type?  
If it's a grid, it will not work.  You can only point to a Details of Form view type.

I tested with a connected field and it works as expected.

**Can you share a screen grab?**  

If not, is it due to confidentiality?  I am working on a `_obf `keyword that **obfuscates** all desired data in a page, so we can share our screens without compromising sensitive data.

You can also grab the Builder's Page with Sample Data like this:

![image](https://github.com/cortexrd/Knack-Toolkit-Library/assets/42851359/14f5a97a-a17b-4de9-99ef-175ee344614f)


ND

**Reply by jbk68** [reply]
Hey Norm, thanks for the reply.  Yeah I think I mis-spoke. Your right my code does say hide rich text when NO, but when I change the field values for field 376 from NO to YES, it should show the rich text view based on the code, which it's not.  Rich text view is being shown regardless of the value of field 376 (YES or NO).  Ever since I removed the previous code, it's no longer toggling from being hidden or shown. ???

View_120 is most definitely a grid view which contains a table with field_376 which I'm referencing in the code that is placed in the rich view (view_253)  **_hv=ktlCond, is, NO, field_376, view_120**

I have no problem showing you a screen shot, but as you said, the code will not work with a grid view and that is all that is on that scene except for the rich text I'm trying to hide or show based on value of field 376 on grid view 120.

Thanks again... JON

**Reply by cortexrd** [reply]
Your comparison field is field_376 in grid view_120.

What happens if you have **10** rows, 5 of which are YES and the rest are NO?
What value do you expect the ktlCond to use to hide or show your Rich Text view?  Any random one?  Make an average?  The first one?

This is the problem with grid views and ktlCond: your grid contains several of that same field_376 - one per row.

Unless there's something I'm missing?

Norm

**Reply by jbk68** [reply]
Your not missing anything other than all rows in grid/table are either "YES" or "NO" **NEVER** both at the same time.

With the exception of showing the rich text momentarily when going between view with rich text and others, the below worked prior to 17.0 and or the _hv fix:
_rvr=view_253
_cls=[!ktlDisplayNone], [ktlTarget, view_253]
_cfv=[field_376, is, NO,,,,h], [ktlTarget, $('#view_253')]

I removed the new keyword code _hv=ktlCond, is, NO, field_376, view_120 in the Rich Text view and put the following in view_120 and it works okay with the exception now that the keyword _cls is removed, it doesn't refresh it self when going from hidden to shown.  Works great when going from being shown to hidden (YES to NO):

_rvr=view_253
_cfv=[field_376, is, NO,,,,h], [ktlTarget, $('#view_253')]

I tried using above with keyword _cls and it doesn't work at all.  

Thanks... JON

**Reply by amandajane-mo** [reply]
Hi Jon,

If the value, for field 376, in your grid is always either YES or NO for all records could you not hold that value in a separate connected object, then display that value in a details view? Then reference the details field in your keyword/s.

Amanda

**Reply by jbk68** [reply]
Hey Amada, I literally was working on a this exact solution as we speak. 

Question for you, using KTL, how do you hide a field based on another fields value?  Nothing to do with above issue, something totally different.

**Reply by amandajane-mo** [reply]
I would use _cfv with ktlTarget. Put the _cfv on the view that holds the field you're basing your decision on.

I do it here:

`_cfv=[Discharge Address, not, , , , , r], [ktlTarget, $('#view_9 .field_315')]`

I'm hiding field 315 if Discharge Address is not blank. To hide the whole field and not just the value I had to include the view_id (not exactly sure why).

Though I think, based on the latest ktl updates you could _hf with ktlCond, I've just not had chance to test these new exciting features yet.

Amanda

**Reply by jbk68** [reply]
Got it, will give it a try

Thank you... JON
